crates/miden-lib/src/account/interface/mod.rs

```diff
@@ -257,9 +261,120 @@ impl AccountComponentInterface {
 
         component_interface_vec
     }
+
+    /// Returns the script body that sends notes to the recipients.
+    ///
+    /// Errors if:
+    /// - the interface does not support the generation of the standard `send_note` procedure.
+    /// - the sender of the note isn't the account for which the script is being built.
+    /// - the note created by the faucet doesn't contain exactly one asset.
+    /// - a faucet tries to distribute an asset with a different faucet ID.
+    pub(crate) fn send_note_procedure(
+        &self,
+        sender_account_id: AccountId,
+        notes: &[PartialNote],
+    ) -> Result<String, TransactionScriptBuilderError> {
+        let mut body = String::new();
+
+        for partial_note in notes {
+            if partial_note.metadata().sender() != sender_account_id {
+                return Err(TransactionScriptBuilderError::InvalidSenderAccount(
+                    partial_note.metadata().sender(),
+                ));
+            }
+
+            body.push_str(&format!(
+                "
+                    push.{recipient}
+                    push.{execution_hint}
+                    push.{note_type}
+                    push.{aux}
+                    push.{tag}
+                    ",
+                recipient = prepare_word(&partial_note.recipient_digest()),
+                note_type = Felt::from(partial_note.metadata().note_type()),
+                execution_hint = Felt::from(partial_note.metadata().execution_hint()),
+                aux = partial_note.metadata().aux(),
+                tag = Felt::from(partial_note.metadata().tag()),
+            ));
+
+            match self {
+                AccountComponentInterface::BasicFungibleFaucet => {
+                    if partial_note.assets().num_assets() != 1 {
+                        return Err(TransactionScriptBuilderError::FaucetNoteWithoutAsset);
+                    }
+
+                    // SAFETY: We checked that the note contains exactly one asset
+                    let asset =
+                        partial_note.assets().iter().next().expect("note should contain an asset");
+
+                    if asset.faucet_id_prefix() != sender_account_id.prefix() {
+                        return Err(TransactionScriptBuilderError::InvalidAsset(
+                            asset.faucet_id_prefix(),
+                        ));
+                    }
```

Comment:
Why does the faucet have this restriction (exactly one assset, asset must be from faucet) while the basic wallet does not? I think I'm missing the intended use case.

Reply:
Sending assets for the faucet implies minting them, which in turn implies that the asset should be from itself only.

Reply:
Thanks, got it. I missed that this calls `distribute`.